File: solutions/115.Distinct_Subsequences/AC_dp_n*m.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_dp_n*m.py
# Create Date: 2015-03-11 11:35:56
# Usage:       AC_dp_n*m.py 
# Descripton:  


class Solution:
    # @return an integer
    def numDistinct(self, S, T):
        n = len(S)
        m = len(T)
        dp = [[0] * (m + 1), [0] * (m + 1)]
        dp[0][0] = 1
        dp[1][0] = 1
        cur = 0

        for i in xrange(1, n + 1):
            cur = 1 - cur
            for j in xrange(1, m + 1):
                if S[i - 1] == T[j - 1]:
                    dp[cur][j] = dp[1-cur][j] + dp[1-cur][j-1]
                else:
                    dp[cur][j] = dp[1-cur][j]

        return dp[cur][m]

# debug
s = Solution()
print s.numDistinct('ccc', 'c')
print s.numDistinct('', 'a')
print s.numDistinct('rabbbit', 'rabbit')

File: solutions/115.Distinct_Subsequences/AC_dp_n*m.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_dp_n*m.cpp
*  Create Date: 2015-03-11 10:53:52
*  Descripton:  dp[i][j]: result of sub-question(S[1...i], T[1..j])
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
public:
    int numDistinct(string S, string T) {
        int n = S.length();
        int m = T.length();

        vector<vector<int> > dp(n + 1, vector<int>(m + 1));

        // init
        for (int i = 0; i <= n; ++i) {
            dp[i][0] = 1;
        }
        
        for (int i = 1; i <= n; ++i) {
            for (int j = 1; j <= m; ++j) {
                if (S[i - 1] == T[j - 1])
                    dp[i][j] = dp[i - 1][j] + dp[i - 1][j - 1];
                else
                    dp[i][j] = dp[i - 1][j];
            }
        }

        return dp[n][m];
    }
};

int main() {

    return 0;
}

